pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -1751,6 +1751,103 @@ impl Timeline {
         Ok(())
     }
 
+    /// Split a gc-compaction job into multiple compaction jobs. Optimally, this function should return a vector of
+    /// `GcCompactionJobDesc`. But we want to keep it simple on the tenant scheduling side without exposing too much
+    /// ad-hoc information about gc compaction itself.
+    pub(crate) async fn gc_compaction_split_jobs(
+        self: &Arc<Self>,
+        options: CompactOptions,
+    ) -> anyhow::Result<Vec<CompactOptions>> {
+        if !options.sub_compaction {
+            return Ok(vec![options]);
+        }
+        let compact_range = options.compact_range.clone().unwrap_or(CompactRange {
+            start: Key::MIN,
+            end: Key::MAX,
+        });
+        let compact_lsn = if let Some(compact_lsn) = options.compact_below_lsn {
+            compact_lsn
+        } else {
+            let gc_info = self.gc_info.read().unwrap();
+            gc_info.cutoffs.select_min() // use the real gc cutoff
+        };
+        let mut compact_jobs = Vec::new();
+        // For now, we simply use the key partitioning information; we should do a more fine-grained partitioning
+        // by estimating the amount of files read for a compaction job. We should also partition on LSN.
+        let partition = self.partitioning.lock().await;
+        let ((dense_ks, sparse_ks), _) = &*partition;
```

Comment:
Why is it a smart choice to use the `self.partitioning` ?

It has been computed at a more-or-less-recent `last_record_lsn` . But the keyspace at our `compact_lsn`, which may be several days in the past, might have looked a lot different.
Why not recompute the partitioning right here? Because it takes a long time => doesn't fit into mgmt api request latency budget?

Reply:
The partition doesn't need to be very accurate and it's not a hard constraint. We will compute our own boundary based on this piece of information. Plus I think scheduling the compaction job should be a relatively fast process.